### ISSUE TITLE
2FA support for Roundcube

### ISSUE THREAD

**Issue by aminb** [kaiyou/freeposte.io#67]
It'd be great to have tow-factor authentication (2FA) in Roundcube. We can use the [`twofactor_gauthenticator`](https://plugins.roundcube.net/packages/alexandregz/twofactor_gauthenticator) plugin for this.

**Comment by kaiyou** [comment]
If there is a nice and simple way to insert plugins into Roundcube, I see no reason not to add some in the Dockerfile, maybe add some configuration variables for sensitive plugins that could be disabled.

However, I cannot see how twofactor authentication will improbe security for a mail server that exposes authentication through IMAP and Submission addiationally to Roundcube.

**Comment by aminb** [comment]
> If there is a nice and simple way to insert plugins into Roundcube, I see no reason not to add some in the Dockerfile, maybe add some configuration variables for sensitive plugins that could be disabled.

Agreed. Installing the 2FA plugin I mentioned above should be pretty [straightforward](https://github.com/alexandregz/twofactor_gauthenticator#installation). The installation of carddav plugin however is more involved, and using Composer seems to be the easiest way to do it. I recommend taking a look at sovereign's [roundcube role](https://github.com/sovereign/sovereign/tree/master/roles/webmail) and the related last few [commits](https://github.com/sovereign/sovereign/commits/master/roles/webmail).

> However, I cannot see how twofactor authentication will improbe security for a mail server that exposes authentication through IMAP and Submission addiationally to Roundcube.

Good point. I guess in order to have _effective_ 2FA, we'd also have to use "application-specific passwords". I found a [few](http://www.justinbuchanan.com/blog/post/2012/12/02/Application-Specific-Passwords-for-Dovecot-and-Postfix) blog [posts](https://www.dgsiegel.net/news/2013_05_21-application_specific_passwords_for_dovecot) and [links](http://serverfault.com/a/726498) looking [around](https://www.happyassassin.net/2014/08/26/adding-application-specific-passwords-to-dovecot-when-using-system-user-accounts/). I haven't yet taken a close look at any of them though.

**Comment by kaiyou** [comment]
Application specific password would be nice. I will not have time to implement any of it in the next couple of weeks, but feel free to start working on a fork. It is definitely the first step towards 2FA for Web applications.

**Comment by aminb** [comment]
Unfortunately I'm quite busy with school and work myself, but if I find any free time I'll try to give it a shot.

**Comment by kaiyou** [comment]
One excellent article on the matter, for further reference: https://umrysh.com/time-based-one-time-passwords-for-email/

**Comment by kaiyou** [comment]
A friend did work and suggested a generic way of tackling the 2FA issue:
- have a 2FA API with the following features:
 - list current authentication attempts that require 2FA
 - confirm a given authentication attempt
 - configure 2FA
- provide modular 2FA, with for instance simple click confirmation (would not protect against stolen credentials), TOTP, and why not something more complex in the future
- provide a Webpage for accessing the 2FA endpoint, and why not other clients

The reasoning is that neither SMTP nor IMAP offer proper bindings for implementing MFA. Also, Mailu the project is to centralize authentication on a given container to minimize complexity while allowing for more sophisticated authentication schemes (LDAP, etc.)

The solution would consist in a blocking authentication (with a decent timeout) and a mechanism to unblock by confirming or denying authentication from the API endpoints. It does sound like a poor solution but it is the only way we thought of. Also, we could produce a couple guidelines for setting up MFA in any generic application that does not support anything native.

**Comment by kaiyou** [comment]
This will be related to #272 and could actually be implemented then if we succeed.

**Comment by agent126** [comment]
Hello, i have RoundCube webmail 1.6.1. This [Plugin](https://github.com/alexandregz/twofactor_gauthenticator), don*t work, does not appear on the menu. Please help
